chromium_src/third_party/blink/renderer/modules/webaudio/analyser_node.cc

```diff
@@ -0,0 +1,121 @@
+
+#include "third_party/blink/renderer/modules/webaudio/analyser_node.h"
+
+#include "base/strings/string_number_conversions.h"
+#include "crypto/hmac.h"
+#include "net/base/registry_controlled_domains/registry_controlled_domain.h"
+#include "third_party/blink/public/platform/web_content_settings_client.h"
+#include "third_party/blink/renderer/core/dom/document.h"
+#include "third_party/blink/renderer/core/frame/local_dom_window.h"
+#include "third_party/blink/renderer/core/frame/local_frame.h"
+#include "third_party/blink/renderer/platform/bindings/script_state.h"
+#include "third_party/blink/renderer/platform/heap/handle.h"
+#include "third_party/blink/renderer/platform/supplementable.h"
+
+using blink::Document;
+using blink::GarbageCollected;
+using blink::HeapObjectHeader;
+using blink::MakeGarbageCollected;
+using blink::Supplement;
+using blink::TraceDescriptor;
+using blink::TraceTrait;
+
+namespace brave {
+
+class BraveSessionCache final : public GarbageCollected<BraveSessionCache>,
+                                public Supplement<Document> {
+  USING_GARBAGE_COLLECTED_MIXIN(BraveSessionCache);
+
+ public:
+  static const char kSupplementName[];
+
+  explicit BraveSessionCache(Document&);
+  virtual ~BraveSessionCache() = default;
+
+  static BraveSessionCache& From(Document&);
+
+  void SetFudgeFactor(double fudge_factor);
+  double GetFudgeFactor();
+  bool IsInitialized();
+
+ private:
+  double fudge_factor_;
+  bool initialized_;
+};
+
+const char BraveSessionCache::kSupplementName[] = "BraveSessionCache";
+
+BraveSessionCache::BraveSessionCache(Document& document)
+    : Supplement<Document>(document), fudge_factor_(0.0), initialized_(false) {}
+
+BraveSessionCache& BraveSessionCache::From(Document& document) {
+  BraveSessionCache* cache =
+      Supplement<Document>::From<BraveSessionCache>(document);
+  if (!cache) {
+    cache = MakeGarbageCollected<BraveSessionCache>(document);
+    ProvideTo(document, cache);
+  }
+  return *cache;
+}
+
+void BraveSessionCache::SetFudgeFactor(double fudge_factor) {
+  fudge_factor_ = fudge_factor;
+  initialized_ = true;
+}
+
+double BraveSessionCache::GetFudgeFactor() {
+  CHECK(initialized_);
+  return fudge_factor_;
+}
+
+bool BraveSessionCache::IsInitialized() {
+  return initialized_;
+}
+
+double GetFudgeFactor(Document* document) {
+  CHECK(document);
+  double fudge_factor;
+  if (BraveSessionCache::From(*document).IsInitialized()) {
+    fudge_factor = BraveSessionCache::From(*document).GetFudgeFactor();
+  } else {
+    base::StringPiece host =
+        base::StringPiece(document->TopFrameOrigin()->ToUrlOrigin().host());
+    std::string domain = net::registry_controlled_domains::GetDomainAndRegistry(
+        host, net::registry_controlled_domains::INCLUDE_PRIVATE_REGISTRIES);
+    crypto::HMAC h(crypto::HMAC::SHA256);
+    uint64_t key;
+    base::StringToUint64(
+        base::CommandLine::ForCurrentProcess()->GetSwitchValueASCII("brave_session_token"), &key);
```

Comment:
what does `StringToUint64` do if the switch value is empty or not a string for some reason?

Reply:
similar code in Chromium is preceded by a DCHECK on `HasSwitch`, so I added that.